libs/shacl/src/lib/services/shacl-shapes-graph-helper.service.ts

```diff
@@ -0,0 +1,400 @@
+  private helperCache: TtlCache<ExpandedJsonLdContainer, ShaclShapesGraphHelper> = inject(TtlCacheFactory).create({
+    ttl: 5000,
+  });
+
+  create(shapesGraph: ExpandedJsonLdContainer, purpose: Many<HanamiPurpose>): ShaclShapesGraphHelper {
+    // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
+    if (this.helperCache.has(shapesGraph)) return this.helperCache.get(shapesGraph)!;
+    const helper = new ShaclShapesGraphHelper(this.jsonLdService, shapesGraph, purpose);
+    this.helperCache.set(shapesGraph, helper);
+    return helper;
+  }
+}
+
+export class ShaclShapesGraphHelper {
+  nodeShapeUris: string[] = [];
+  propertyShapeUris: string[] = [];
+  propertyShapeUrisByPath: Record<string, string[] | undefined> = {};
+  nodeShapeUrisByTargetClass: Record<string, string[] | undefined> = {};
+  nodeShapeUrisByTargetNode: Record<string, string[] | undefined> = {};
+  nodeShapeUrisByTargetObjectsOf: Record<string, string[] | undefined> = {};
+  hasTargetObjectsOf = false;
+  propertyPathWithShNode: string[] = [];
+  hasShNode = false;
+  nodeShapeUrisByTargetSubjectsOf: Record<string, string[] | undefined> = {};
+  classUrisAndAncestors: Record<string, string[]> = {};
+  private memoizer = new Memoizer();
+
+  constructor(
+    private readonly jsonLdService: JsonLdService,
+    public readonly shapesGraph: ExpandedJsonLdContainer,
+    public readonly purpose: Many<HanamiPurpose>
+  ) {
+    this.initMemoized();
+    this.buildCache();
+  }
+
+  isAttribute(nodeShapeUri: string, propertyPath: string): boolean {
+    const property = this.getPropertyShape(nodeShapeUri, propertyPath);
+    if (!property) return false;
+    return this.isPropertyAttribute(property);
+  }
+
+  isPropertyAttribute(propertyShape: ShPropertyShape): boolean {
+    return (
+      propertyShape[SH.datatype] != null ||
+      this.isUntypedIri(propertyShape) ||
+      !!this.getConsolidatedPropertyShapesSortedBySpecificity(propertyShape['@id'])?.some(subProperty => {
+        if (subProperty['@id'] === propertyShape['@id']) return false;
+        return this.isPropertyAttribute(subProperty);
+      })
+    );
+  }
+
+  isReference(nodeShapeUri: string, propertyPath: string): boolean {
+    const property = this.getPropertyShape(nodeShapeUri, propertyPath);
+    return property ? this.isPropertyReference(property) : false;
+  }
+
+  isPropertyReference(propertyShape: ShPropertyShape): boolean {
+    return (
+      !!getAllValues(propertyShape?.[SH.class], this.shapesGraph).length ||
+      !!propertyShape?.[HANAMI.selection] ||
+      getConcreteNodeKindsOfPropertyShape(propertyShape, this.shapesGraph).includes(SH.BlankNode) ||
+      !!this.getConsolidatedPropertyShapesSortedBySpecificity(propertyShape['@id']).some(subProperty => {
+        if (subProperty['@id'] === propertyShape['@id']) return false;
+        return this.isPropertyReference(subProperty);
+      })
+    );
+  }
+
+  isSingle(nodeShapeUri: string, propertyPath: string): boolean {
```

Comment:
we already have this defined in the [shacl-data-graph-helper.service.ts](https://github.com/cognizone/ng-cognizone/pull/63/files#diff-72f0265abd9a6a5085d564f94c95c131e91c224e621f3d652a7f8db406608e22) 🤔  (do we need to have these methods here again or can they be removed?)

Reply:
I know it looks A LOT like being the same method, but one uses a `nodeUri`, the other a `nodeShapeUri`, which are 2 different things ^^'